ci-build.sh

```diff
@@ -26,6 +28,11 @@ while [[ -n "$1" ]]; do
             export TEMP_POSTGRES=1
             echo Using temp database
             ;;
+    "--check-test-tx-meta")
+            # When testing against a TxMeta baseline, we turn off instrumentation
+            # to have a faster build & test cycle.
+            INSTRUMENTATION_CONFIG=""
```

Comment:
re: `INSTRUMENTATION_CONFIG` approach.
Concern I have here is that we're doubling the size of caches (as we build asan and non asan builds) and we're going to hit the global Github limit between the `auto` branch and PRs.

I don't know if having that leg faster makes any difference as we're limited by the slowest build leg anyways - so maybe we should just use the same build flavors?

An alterrnative would be to do something like what we used to do in Travis: split the "build" step from "test".

* first we'd perform the "build" step, that would just build without running tests, and would populate the cache
* "test" steps would build again, but with caches the time spent building was very low

Reply:
Hmm. Looking into this: github says its cache limit is 5GB. If I do a clean build with the CFLAGS in ci-build.sh (`-g1 -O2`), ccache says I use 1GB.

We are currently doing 4 builds but the protocol dimension appears to only affect a single file (Config.cpp) so I suspect we really only have 1 dimension (toolchain) to work with and should eliminate the protocol dimension -- a cache miss on `Config.cpp` is harmless, so long as the presence of the additional preprocessor symbol `ENABLE_NEXT_PROTOCOL_VERSION_UNSAFE_FOR_PRODUCTION` doesn't cause ccache misses across the board (I will check on this and propose a patch to work around it if so).

Given that, I think we have room to have a 2-dimensional / 4-build cache: {gcc,clang} and {instrumentation,none}.

Another option I'd be happy to explore is turning off `-g1` from one or more of the builds. It's not clear to me that we get a lot of value out of the debuginfo _in CI_ (if there's a failure we tend to diagnose locally) and a no-debuginfo build only uses 400MB.

If you prefer to just press ahead on this and get it working as quickly as possible, I concur that recycling the existing cache / not disabling instrumentation on this dimension is the simplest change. Happy to explore any of the options listed above.

Reply:
> If I do a clean build with the CFLAGS in ci-build.sh (`-g1 -O2`), ccache says I use 1GB.

That number doesn't sound right to me. I was testing this recently with regard to setting the cache size correctly, and I recall that 1GB is approximately the size of the cache in the absence of any compression flags. But ci-build.sh does set some compression flags (`CCACHE_COMPRESS=true` and `CCACHE_COMPRESSLEVEL=9`), and in that case the result should be 300-400 MB based on my experience. The 300-400 MB number is also corroborated by the fact that the max cache size of 500 MB does actually work effectively.

Regardless of the above, it seems highly desirable if we can avoid caching on the protocol dimension so I'm interested in that proposal.

Reply:
Followup: I did some testing with ccache and it appears that the change in command-line defines (that is: adding `-DENABLE_NEXT_PROTOCOL_VERSION_UNSAFE_FOR_PRODUCTION`) does not actually cause a cache miss.

It does cause a _direct mode_ cache miss (where ccache merely hashes the command line and inputs), but when that happens ccache falls back to _preprocessor mode_ where it runs `cpp` and hashes the resulting output. Since this symbol has no effect on the majority of the files, they still get a cache hit at this point.

However, I see that there is a 500MB cache limit (as Jon mentions) and I believe this has to do with the desire to have multiple caches in the github store, in order to support _updating_ the cache using hourly timestamps and cache-key fallbacks.

Based on these observations, I suggest (and will modify the PR to conform to) the following:

  - Remove the protocol dimension of the cache key
  - Remove the scenario dimension of the cache key
  - Remove the code that disables instrumentation in TxMeta-checking scenario

In other words, only a toolchain dimension will remain in the cache key, and all builds will have instrumentation enabled.

We may still want to remove debuginfo to give us more breathing room wrt. the cache limit, but for now I will leave it enabled. I measured with `CFLAGS=-g1 -O2` and `CCACHE_COMPRESSLEVEL=9` and it appears to only be 185MB -- compression helps a lot! -- so we're probably ok leaving it on.